### ISSUE TITLE
`openssl cms -sign/-verify` do not handle binary input properly (including DER)

### ISSUE THREAD

**Issue by ximaera** [openssl/openssl#8940]
The binary input is copied to signedData.encapContentInfo.eContent only up to the first octet equal to zero, the rest gets lost. The `-binary` flag has no effect on this.

Sample command line:

    ```
    $ openssl version
    OpenSSL 1.0.2r  26 Feb 2019
    $ alias _dd="dd if=/dev/random count=1 2>/dev/null"
    $ alias _sign="openssl cms -sign -md sha256 -signer cert.pem -inkey key.pem"
    $ alias _verify="openssl cms -verify -CAfile ca.pem 2>/dev/null"
    $ for _i in {1..3}; do _dd bs=512 | _sign | _verify | wc -c; done
         351
          36
         244
    $
    ```

**Comment by beldmit** [comment]
DESCRIPTION
       The cms command handles S/MIME v3.1 mail. It can encrypt, decrypt, sign
       and verify, compress and uncompress S/MIME messages.

**Comment by ximaera** [comment]
Well,

1. The `cms(1)` manpage states that `openssl cms` might as well be used with DER-encoded messages, and the `-encrypt` command actually produces such messages if invoked with an `-outform DER` argument. A DER-encoded message would always contain zeros.

As a matter of fact, the same behaviour reproduces if you supply `cms -sign` with an output of `cms -encrypt -outform DER`.

2. Even if `cms -sign` could only process S/MIME format, the right way to implement that is probably to check _if_ the input is, in fact, S/MIME, and to fail otherwise. Not to silently generate a truncated response which would then result in _cryptic_ error messages when supplied as an input to `cms -verify` (as it is now).

**Comment by beldmit** [comment]
```       -inform SMIME|PEM|DER
           this specifies the input format for the CMS structure. The default
           is SMIME which reads an S/MIME format message. PEM and DER format
           change this to expect PEM and DER format CMS structures instead.
           This currently only affects the input format of the CMS structure,
           if no CMS structure is being input (for example with -encrypt or
           -sign) this option has no effect.
```
I agree that verifying format will be a good idea.

**Comment by ximaera** [comment]
Fine. I will leave the issue open for now in case checking the input would be treated necessary later.

**Comment by DDvO** [comment]
The error report itself is flawed.
Piping random binary data, as done in the given example easily yields to truncation already the the shell level.
Therefore already the input to `openssl cms` went wrong.

**Comment by DDvO** [comment]
Here is a better example:
```
dd 2>/dev/null if=/dev/zero count=1 bs=512 of=zeros.bin
openssl cms -sign -md sha256 -signer test/certs/p384-root.pem -inkey test/certs/p384-root-key.pem -in zeros.bin -out zeros.signed.bin
openssl cms -verify -CAfile test/certs/p384-root.pem 2>/dev/null -in zeros.signed.bin -out zeros.verified.bin
ls -l zeros.*
```
which yields
```
-rw-r--r-- 1 david david  512 Sep 22 12:27 zeros.bin
-rw-r--r-- 1 david david 1689 Sep 22 12:27 zeros.signed.bin
-rw-r--r-- 1 david david    0 Sep 22 12:27 zeros.verified.bin
```
So what actually gets truncated by OpenSSL is the binary output of `openssl cms -verify` as soon as a zero byte is present.
```

**Comment by DDvO** [comment]
It turns out that this 'bug' is due to insufficient usage of CMS CLI options,
also due to poor documentation and user guidance regarding binary input.

One needs to add `binary -outform der` or `-binary -outform pem` to the sign call
and `-binary -inform der -content text_zeros.bin` or ` -binary -inform der -content text_zeros.bin` to the verify call:
```
openssl cms -sign -md sha256 -signer test/certs/p384-root.pem -inkey test/certs/p384-root-key.pem -in text_zeros.bin -out text_zeros.signed.bin -binary -outform der
openssl cms -verify -CAfile test/certs/p384-root.pem -in text_zeros.signed.bin -out text_zeros.verified.bin -binary -inform der -content text_zeros.bin
```
Then it works as expected:
```
Verification successful
-rw-r--r-- 1 david david   10 Sep 22 15:09 text_zeros.bin
-rw-r--r-- 1 david david 1269 Sep 22 15:09 text_zeros.signed.bin
-rw-r--r-- 1 david david   10 Sep 22 15:09 text_zeros.verified.bin
```

**Comment by DDvO** [comment]
Actually it should be possible to use binary input simply with `-binary` but without `-outform der` or `-outform pem`.
I'll provide a fix for that, 
also improving the documentation and adding a warning in case the `-binary` option should have been used.

**Comment by DDvO** [comment]
Fix provided in #12959.
Thanks @ximaera for reporting this issue and leaving it open until this is (hopefully) sorted out now.

**Comment by DDvO** [comment]
Should finally be fixed now :smile:
